File: iosApp/iosApp/viewmodel/ExpertDocumentUploadViewModel.swift
import Foundation
import SwiftUI
import FirebaseStorage
import FirebaseAuth
import Shared
import UniformTypeIdentifiers

@MainActor
class ExpertDocumentUploadViewModel: ObservableObject {
    @Published var uploadedDocumentURL: String?
    @Published var uploadedFileName: String?
    @Published var isUploading: Bool = false
    @Published var uploadProgress: Double = 0.0
    @Published var error: String?
    
    private let storageBucket = "gs://kissangram-19531.firebasestorage.app"
    
    private let prefs = IOSPreferencesRepository()
    private lazy var authRepo = IOSAuthRepository(preferencesRepository: prefs)
    private lazy var userRepo = FirestoreUserRepository(authRepository: authRepo)
    private lazy var createUserProfileUseCase = CreateUserProfileUseCase(
        authRepository: authRepo,
        preferencesRepository: prefs,
        userRepository: userRepo
    )
    
    func uploadDocument(fileURL: URL, userId: String) async {
        await MainActor.run {
            isUploading = true
            uploadProgress = 0.0
            error = nil
        }
        
        do {
            let fileName = fileURL.lastPathComponent
            let fileExtension = fileURL.pathExtension
            
            let storage = Storage.storage(url: storageBucket)
            let documentRef = storage.reference().child("expert-documents/\(userId)/\(UUID().uuidString).\(fileExtension)")
            
            let fileData = try Data(contentsOf: fileURL)
            
            let metadata = StorageMetadata()
            metadata.contentType = getContentType(for: fileExtension)
            
            let uploadTask = documentRef.putData(fileData, metadata: metadata)
            
            uploadTask.observe(StorageTaskStatus.progress) { [weak self] (snapshot: StorageTaskSnapshot) in
                guard let progress = snapshot.progress else { return }
                let percentComplete = Double(progress.completedUnitCount) / Double(progress.totalUnitCount)
                Task { @MainActor in
                    self?.uploadProgress = percentComplete
                }
            }
            
            try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
                var hasResumed = false
                
                uploadTask.observe(StorageTaskStatus.success) { (_: StorageTaskSnapshot) in
                    if !hasResumed {
                        hasResumed = true
                        continuation.resume()
                    }
                }
                
                uploadTask.observe(StorageTaskStatus.failure) { (snapshot: StorageTaskSnapshot) in
                    if !hasResumed, let err = snapshot.error {
                        hasResumed = true
                        continuation.resume(throwing: err)
                    }
                }
            }
            
            let downloadURL = try await documentRef.downloadURL()
            
            await MainActor.run {
                self.uploadedDocumentURL = downloadURL.absoluteString
                self.uploadedFileName = fileName
                self.isUploading = false
                self.uploadProgress = 1.0
            }
        } catch {
            await MainActor.run {
                self.error = error.localizedDescription
                self.isUploading = false
            }
        }
    }
    
    private func getContentType(for fileExtension: String) -> String {
        switch fileExtension.lowercased() {
        case "pdf": return "application/pdf"
        case "jpg", "jpeg": return "image/jpeg"
        case "png": return "image/png"
        case "doc": return "application/msword"
        case "docx": return "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
        default: return "application/octet-stream"
        }
    }
    
    func completeSetup(onSuccess: @escaping () -> Void, onError: @escaping (String) -> Void = { _ in }) {
        Task {
            do {
                let docUrl = uploadedDocumentURL
                try await createUserProfileUseCase.invoke(
                    role: .expert,
                    verificationDocUrl: docUrl,
                    verificationStatus: docUrl != nil ? .pending : .unverified
                )
                try await prefs.setAuthCompleted()
                await MainActor.run {
                    error = nil
                    onSuccess()
                }
            } catch {
                await MainActor.run {
                    self.error = error.localizedDescription
                    onError(error.localizedDescription)
                }
            }
        }
    }
    
    func skipVerification(onSuccess: @escaping () -> Void, onError: @escaping (String) -> Void = { _ in }) {
        Task {
            do {
                try await createUserProfileUseCase.invoke(
                    role: .expert,
                    verificationDocUrl: nil,
                    verificationStatus: .unverified
                )
                try await prefs.setAuthCompleted()
                await MainActor.run {
                    error = nil
                    onSuccess()
                }
            } catch {
                await MainActor.run {
                    self.error = error.localizedDescription
                    onError(error.localizedDescription)
                }
            }
        }
    }
}

File: composeApp/src/androidMain/kotlin/com/kissangram/viewmodel/ExpertDocumentUploadViewModel.kt
package com.kissangram.viewmodel

import android.app.Application
import android.content.Context
import android.net.Uri
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.storage.FirebaseStorage
import com.kissangram.model.VerificationStatus
import com.kissangram.model.UserRole
import com.kissangram.repository.AndroidAuthRepository
import com.kissangram.repository.AndroidPreferencesRepository
import com.kissangram.repository.FirestoreUserRepository
import com.kissangram.usecase.CreateUserProfileUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await

data class ExpertDocumentUploadUiState(
    val uploadedDocumentURL: String? = null,
    val uploadedFileName: String? = null,
    val isUploading: Boolean = false,
    val uploadProgress: Float = 0f,
    val error: String? = null
)

class ExpertDocumentUploadViewModel(
    application: Application
) : AndroidViewModel(application) {

    private val appContext = application.applicationContext
    private val prefs = AndroidPreferencesRepository(appContext)
    private val authRepository = AndroidAuthRepository(
        context = appContext,
        activity = null,
        preferencesRepository = prefs
    )
    private val userRepository = FirestoreUserRepository(authRepository = authRepository)
    private val createUserProfileUseCase = CreateUserProfileUseCase(
        authRepository = authRepository,
        preferencesRepository = prefs,
        userRepository = userRepository
    )

    private val _uiState = MutableStateFlow(ExpertDocumentUploadUiState())
    val uiState: StateFlow<ExpertDocumentUploadUiState> = _uiState.asStateFlow()

    private val storage = FirebaseStorage.getInstance("gs://kissangram-19531.firebasestorage.app")

    fun uploadDocument(uri: Uri, context: Context, userId: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(
                isUploading = true,
                uploadProgress = 0f,
                error = null
            )

            try {
                val fileName = getFileName(uri, context) ?: "document_${System.currentTimeMillis()}"
                val fileExtension = fileName.substringAfterLast('.', "")
                val storageRef = storage.reference
                val documentRef = storageRef.child("expert-documents/$userId/${System.currentTimeMillis()}.$fileExtension")
                val metadata = com.google.firebase.storage.StorageMetadata.Builder()
                    .setContentType(getContentType(fileExtension))
                    .build()
                val uploadTask = documentRef.putFile(uri, metadata)

                uploadTask.addOnProgressListener { taskSnapshot ->
                    val progress = (100.0 * taskSnapshot.bytesTransferred / taskSnapshot.totalByteCount).toFloat()
                    _uiState.value = _uiState.value.copy(uploadProgress = progress / 100f)
                }

                val task = uploadTask.await()
                val downloadUrl = task.storage.downloadUrl.await()

                _uiState.value = _uiState.value.copy(
                    uploadedDocumentURL = downloadUrl.toString(),
                    uploadedFileName = fileName,
                    isUploading = false,
                    uploadProgress = 1f
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    error = e.message ?: "Upload failed",
                    isUploading = false
                )
            }
        }
    }

    private fun getFileName(uri: Uri, context: Context): String? {
        var result: String? = null
        val cursor = context.contentResolver.query(uri, null, null, null, null)
        cursor?.use {
            if (it.moveToFirst()) {
                val nameIndex = it.getColumnIndex(android.provider.OpenableColumns.DISPLAY_NAME)
                if (nameIndex != -1) {
                    result = it.getString(nameIndex)
                }
            }
        }
        return result
    }

    private fun getContentType(fileExtension: String): String {
        return when (fileExtension.lowercase()) {
            "pdf" -> "application/pdf"
            "jpg", "jpeg" -> "image/jpeg"
            "png" -> "image/png"
            "doc" -> "application/msword"
            "docx" -> "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
            else -> "application/octet-stream"
        }
    }

    fun completeSetup(onSuccess: () -> Unit, onError: (String) -> Unit = {}) {
        viewModelScope.launch {
            try {
                val docUrl = _uiState.value.uploadedDocumentURL
                createUserProfileUseCase(
                    role = UserRole.EXPERT,
                    verificationDocUrl = docUrl,
                    verificationStatus = if (docUrl != null) VerificationStatus.PENDING else VerificationStatus.UNVERIFIED
                )
                prefs.setAuthCompleted()
                _uiState.value = _uiState.value.copy(error = null)
                onSuccess()
            } catch (e: Exception) {
                val msg = e.message ?: "Failed to save"
                _uiState.value = _uiState.value.copy(error = msg)
                onError(msg)
            }
        }
    }

    fun skipVerification(onSuccess: () -> Unit, onError: (String) -> Unit = {}) {
        viewModelScope.launch {
            try {
                createUserProfileUseCase(
                    role = UserRole.EXPERT,
                    verificationDocUrl = null,
                    verificationStatus = VerificationStatus.UNVERIFIED
                )
                prefs.setAuthCompleted()
                _uiState.value = _uiState.value.copy(error = null)
                onSuccess()
            } catch (e: Exception) {
                val msg = e.message ?: "Failed to save"
                _uiState.value = _uiState.value.copy(error = msg)
                onError(msg)
            }
        }
    }
}
